java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/NbProtocolServer.java

```diff
@@ -349,6 +352,32 @@ public NbProtocolServer(DebugAdapterContext context) {
         return future;
     }
 
+    @Override
+    @NbBundle.Messages({"MSG_FrameRestartUnsupported=Restart of frames is not supported.",
+                        "# {0} - frame pop error message",
+                        "MSG_FrameRestartFailed=Unable to restart frame: {0}"})
+    public CompletableFuture<Void> restartFrame(RestartFrameArguments args) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        NbFrame stackFrame = (NbFrame) context.getThreadsProvider().getThreadObjects().getObject(args.getFrameId());
+        String popError = null;
+        if (stackFrame != null) {
+            try {
+                stackFrame.getDVFrame().popOff();
+                ActionsManager am = DebuggerManager.getDebuggerManager().getCurrentEngine().getActionsManager();
+                am.doAction("stepInto");
+                future.complete(null);
+            } catch (UnsupportedOperationException ex) {
+                popError = Bundle.MSG_FrameRestartUnsupported();
+            } catch (DebuggingView.PopException ex) {
+                popError = Bundle.MSG_FrameRestartFailed(ex.getLocalizedMessage());
```

Comment:
Please note that `PopException` is final and doesn't override `getLocalizedMessage()` - e.g. the value is always going to be same as `getMessage()`.